content/docs/rfcs/32/README.md

```diff
@@ -0,0 +1,554 @@
+* `identity_secret_hash`
+* `path_elements` - rln membership proof component
+* `identity_path_index` - rln membership proof component
+
+**Outputs**
+- `y`
+- `root` - the rln membership tree root
+- `internal_nullifier`
+
+### Hash function
+
+Canonical [Poseidon hash implementation](https://eprint.iacr.org/2019/458.pdf) is used, 
+as implemented in the [circomlib library](https://github.com/iden3/circomlib/blob/master/circuits/poseidon.circom), according to the Poseidon paper.
+This Poseidon hash version (canonical implementation) is uses the following parameters:
+
+- `t`: 3
+- `RF`: 8
+- `RP`: 57
+
+### Membership implementation
+
+For a valid signal, a user's `identity_commitment` (more on identity commitments below) must exist in identity membership tree.
+Membership is proven by providing a membership proof (witness). 
+The fields from the membership proof required for the verification are: 
+`path_elements` and `identity_path_index`.
+
+[IncrementalQuinTree](https://github.com/appliedzkp/incrementalquintree) algorithm is used for constructing the Membership merkle tree. 
+The circuits are reused from this repository. 
+You can find out more details about the IncrementalQuinTree algorithm [here](https://ethresear.ch/t/gas-and-circuit-constraint-benchmarks-of-binary-and-quinary-incremental-merkle-trees-using-the-poseidon-hash-function/7446).
+
+### Slashing and Shamir's Secret Sharing
+
+Slashing is enabled by using polynomials and [Shamir's Secret sharing](https://en.wikipedia.org/wiki/Shamir%27s_Secret_Sharing). 
+In order to produce a valid proof, `identity_secret_hash` as a private input to the circuit. 
+Then a secret equation is created in the form of:
+
+```
+y = a_0 + x*a_1,
+```
+
+where `a_0` is the `identity_secret_hash` and `a_1 = hash(a_0, external nullifier)`.
+Along with the generated proof, 
+the users need to provide a (x, y) share which satisfies the line equation, 
+in order for their proof to be verified. 
+`x` is the hashed signal, while the `y` is the circuit output. 
+With more than one pair of unique shares, anyone can derive `a_0`, the `identity_secret_hash` . 
+Hash of a signal will be evaluation point x. 
+So that a member who sends more that one unique signal per external_nullifier risks their identity secret being revealed.
+
+Note that shares used for different external nullifiers and different RLN apps cannot be used to derive the secret key.
+
+The `rln_identifier` is a random value from a finite field, 
+unique per RLN app, 
+and is used for additional cross-application security - to protect the user secrets being compromised if they use the same credentials accross different RLN apps. 
+If `rln_identifier` is not present, 
+the user uses the same credentials and sends a different message for two different RLN apps using the same `external_nullifier`,
+then their secret key can be revealed. 
+With adding the `rln_identifier` field we obscure the nullifier, 
+so this kind of attack cannot happen. 
```

Comment:
This is true, but that side knowledge would be the internal nullifier itself. Otherwise it is very hard to keep track of whose messages are the targeted user's one. Think of the internal nullifier as a user's epoch based identity. If that is different per app, you can't really find out which messages are which. However if there is an entity that can link messages to the specific user (i.e they identify the user by a certain identifiable property), then this would be problematic but then we have larger problems than the nullifier itself.